File: core/account/tx/handlers/j-events/claim.ts
import type { AccountTx , AccountOutput } from '../../../../types/account';
import type { AccountJClaimSession } from '../../../j-claims/j-claim-session';
import { getAccountPerspective } from '../../../state/perspective';
import { applyAccountJClaimTransition } from '../../../j-claims/j-claim-transition';
import { applyFinalizedAccountJEventsOnView } from './finality';
import {
  getAccountStateDomain,
  requireAccountDeltaTransformerAddress,
  type AccountJurisdictionView,
} from '../../../consensus/helpers';
import { createStructuredLogger, shortHash } from '../../../../support/logger';
import type { ApplyAccountTxResult } from '../../apply-types';
import { accountTxApplied } from '../../apply-result';
import type { AccountDraftReplica } from '../../../state/account-state-draft';

const jEventClaimLog = createStructuredLogger('account.j_event');

export function handleJEventClaim(
  account: AccountDraftReplica,
  accountTx: Extract<AccountTx, { type: 'j_event_claim' }>,
  byLeft: boolean,
  _currentTimestamp: number,
  myEntityId: string,
  candidateEffects: AccountOutput[],
  jurisdictions: AccountJurisdictionView,
  session: AccountJClaimSession,
): ApplyAccountTxResult {
  const { jHeight, jBlockHash } = accountTx.data;
  jEventClaimLog.debug('claim.received', { jHeight, hash: shortHash(jBlockHash), byLeft });
  const { counterparty } = getAccountPerspective(account.state, myEntityId);
  const transition = applyAccountJClaimTransition(
    account.state,
    accountTx,
    byLeft,
    getAccountStateDomain(account.state),
    session,
  );
  if (transition.status === 'pending' || transition.status === 'idempotent' || transition.status === 'stale') {
    account.state.leftPendingJClaims = transition.left;
    account.state.rightPendingJClaims = transition.right;
    return accountTxApplied([transition.status === 'pending'
        ? '📥 J-event claim authenticated and retained'
        : `ℹ️ j_event_claim ${transition.status}`]);
  }

  // `applyAccountTx` already owns the Account transition. Opening another
  // transition here would make draft Patricia collections appear committed and
  // break exact rollback. Finality mutates the existing owned view; rejection
  // discards that one owner at the Account-machine boundary.
  account.state.leftPendingJClaims = transition.left;
  account.state.rightPendingJClaims = transition.right;
  applyFinalizedAccountJEventsOnView(
    account,
    counterparty,
    transition.events,
    requireAccountDeltaTransformerAddress(jurisdictions, account.state),
  );
  account.state.lastFinalizedJHeight = jHeight;
  jEventClaimLog.debug('claim.applied_to_draft', { jHeight });

  const settledTokenId = Number(
    transition.events.find((event) => event.type === 'AccountSettled')?.data?.tokenId ?? 1,
  );
  const delta = account.state.deltas.get(settledTokenId);
  // This function only describes candidate effects; Runtime dispatch happens
  // after Entity commit. Retaining the effect during validation lets the
  // receiver install one verified Account transition instead of replaying it
  // merely to rediscover the same output. Rejected validation drops it.
  const data = {
    entityId: myEntityId,
    accountId: counterparty,
    tokenId: settledTokenId,
    jHeight,
    collateral: String(delta?.collateral ?? 0n),
    ondelta: String(delta?.ondelta ?? 0n),
  };
  candidateEffects.push({
    kind: 'runtimeEvent',
    eventName: 'account_settled_finalized_bilateral',
    data,
  });
  candidateEffects.push({
    kind: 'debug',
    payload: {
      level: 'info',
      code: 'REB_STEP',
      step: 5,
      status: 'ok',
      event: 'account_settled_finalized_bilateral',
      ...data,
    },
  });
  return accountTxApplied(['✅ J-event claim finalized bilaterally']);
}

File: rscore/crates/engine/src/tx/handlers/j_events/claim.rs
use crate::j_claims::{
    AccountJEventClaimAdmissionResult, JClaimStatus, apply_admitted_claim_transition,
    validate_j_event_claim_admission,
};
use crate::tx::apply_types::MutationDecision;
use crate::{AccountRejection, AccountReplica, JEventClaimTx, Side, TransitionError};

use super::finality::apply_finalized_events;

pub(crate) fn apply_j_event_claim(
    replica: &mut AccountReplica,
    tx: &JEventClaimTx,
    proposer: Side,
) -> Result<MutationDecision, TransitionError> {
    let identity = replica.state().identity().clone();
    let left = replica.state().carried().left_pending_j_claims.clone();
    let right = replica.state().carried().right_pending_j_claims.clone();
    let finalized_height = replica.state().last_finalized_j_height();
    let admission = match validate_j_event_claim_admission(&identity, &left, &right, tx)? {
        AccountJEventClaimAdmissionResult::Accepted(admission) => admission,
        AccountJEventClaimAdmissionResult::Rejected(rejection) => {
            return Ok(MutationDecision::rejected(AccountRejection::Validation(
                rejection,
            )));
        }
    };
    let transition = apply_admitted_claim_transition(
        &left,
        &right,
        finalized_height,
        tx,
        proposer == Side::Left,
        replica.state_mut().j_claim_store_mut(),
        *admission,
    )?;
    replica
        .state_mut()
        .set_j_claim_accumulators(transition.left, transition.right);
    let (message, outputs) = match transition.status {
        JClaimStatus::Pending => ("📥 J-event claim authenticated and retained", Vec::new()),
        JClaimStatus::Idempotent => ("ℹ️ j_event_claim idempotent", Vec::new()),
        JClaimStatus::Stale => ("ℹ️ j_event_claim stale", Vec::new()),
        JClaimStatus::Finalized => (
            "✅ J-event claim finalized bilaterally",
            vec![apply_finalized_events(
                replica,
                &transition.events,
                tx.j_height,
            )?],
        ),
    };
    Ok(MutationDecision::with_outputs(
        vec![message.into()],
        outputs,
    ))
}
